File: apps/api/app/models/payment.py
import uuid
from enum import Enum
from datetime import datetime

from sqlalchemy import String, DateTime, Integer, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class PaymentProvider(str, Enum):
    paystack = "paystack"
    flutterwave = "flutterwave"
    stripe = "stripe"


class PaymentStatus(str, Enum):
    pending = "pending"
    completed = "completed"
    failed = "failed"


class Payment(Base):
    __tablename__ = "payments"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: uuid.uuid4().hex[:12])
    event_id: Mapped[str] = mapped_column(String)
    user_id: Mapped[str] = mapped_column(String)
    provider: Mapped[PaymentProvider] = mapped_column(SAEnum(PaymentProvider))
    provider_ref: Mapped[str] = mapped_column(String(255))
    amount: Mapped[int] = mapped_column(Integer)
    currency: Mapped[str] = mapped_column(String(10))
    status: Mapped[PaymentStatus] = mapped_column(SAEnum(PaymentStatus), default=PaymentStatus.pending)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=datetime.utcnow)

File: apps/flutter_app/lib/models/payment.dart
enum PaymentProvider { paystack, flutterwave, stripe }

enum PaymentStatus { pending, completed, failed }

class Payment {
  final String id;
  final String eventId;
  final String userId;
  final PaymentProvider provider;
  final String providerRef;
  final int amount;
  final String currency;
  final PaymentStatus status;
  final DateTime createdAt;

  Payment({
    required this.id,
    required this.eventId,
    required this.userId,
    required this.provider,
    required this.providerRef,
    required this.amount,
    required this.currency,
    required this.status,
    required this.createdAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) => Payment(
    id: json['id'],
    eventId: json['event_id'],
    userId: json['user_id'],
    provider: PaymentProvider.values.byName(json['provider']),
    providerRef: json['provider_ref'],
    amount: json['amount'],
    currency: json['currency'],
    status: PaymentStatus.values.byName(json['status']),
    createdAt: DateTime.parse(json['created_at']),
  );
}
